sdk/src/types/block/output/feature/metadata.rs

```diff
@@ -27,77 +59,78 @@ impl MetadataFeature {
+    ) -> &BTreeMap<BoxedSlicePrefix<u8, MetadataFeatureKeyLength>, BoxedSlicePrefix<u8, MetadataFeatureValueLength>>
+    {
+        self.0.deref()
     }
 }
 
 impl StorageScore for MetadataFeature {}
 
 impl WorkScore for MetadataFeature {}
 
-macro_rules! impl_from_vec {
-    ($type:ty) => {
-        impl TryFrom<$type> for MetadataFeature {
-            type Error = Error;
+impl TryFrom<BTreeMap<Vec<u8>, Vec<u8>>> for MetadataFeature {
+    type Error = Error;
 
-            fn try_from(value: $type) -> Result<Self, Self::Error> {
-                Vec::<u8>::from(value).try_into()
+    fn try_from(data: BTreeMap<Vec<u8>, Vec<u8>>) -> Result<Self, Error> {
+        let mut res = BTreeMap::<
+            BoxedSlicePrefix<u8, MetadataFeatureKeyLength>,
+            BoxedSlicePrefix<u8, MetadataFeatureValueLength>,
+        >::new();
+        for (k, v) in data {
+            if !k.is_ascii() {
+                return Err(Error::NonAsciiMetadataKey(k.to_vec()));
             }
+            res.insert(
+                k.into_boxed_slice()
+                    .try_into()
+                    .map_err(Error::InvalidMetadataFeatureKeyLength)?,
+                v.into_boxed_slice()
+                    .try_into()
+                    .map_err(Error::InvalidMetadataFeatureValueLength)?,
+            );
         }
-    };
-}
-impl_from_vec!(&str);
-impl_from_vec!(String);
-impl_from_vec!(&[u8]);
-
-impl<const N: usize> TryFrom<[u8; N]> for MetadataFeature {
-    type Error = Error;
-
-    fn try_from(value: [u8; N]) -> Result<Self, Self::Error> {
-        value.to_vec().try_into()
-    }
-}
-
-impl TryFrom<Vec<u8>> for MetadataFeature {
-    type Error = Error;
-
-    fn try_from(data: Vec<u8>) -> Result<Self, Error> {
-        data.into_boxed_slice().try_into()
-    }
-}
-
-impl TryFrom<Box<[u8]>> for MetadataFeature {
-    type Error = Error;
-
-    fn try_from(data: Box<[u8]>) -> Result<Self, Error> {
-        data.try_into().map(Self).map_err(Error::InvalidMetadataFeatureLength)
-    }
-}
-
-impl FromStr for MetadataFeature {
-    type Err = Error;
-
-    fn from_str(s: &str) -> Result<Self, Self::Err> {
-        Self::new(prefix_hex::decode::<Vec<u8>>(s).map_err(Error::Hex)?)
+        Ok(Self(res.try_into().map_err(Error::InvalidMetadataFeatureLength)?))
     }
 }
 
 impl core::fmt::Display for MetadataFeature {
     fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
-        write!(f, "{}", prefix_hex::encode(self.data()))
+        write!(
+            f,
+            "{:?}",
+            self.0
+                .keys()
+                // Safe to unwrap, keys must be ascii
+                .map(|k| alloc::str::from_utf8(k).unwrap())
+                .collect::<Vec<&str>>()
+        )
     }
 }
 
 impl core::fmt::Debug for MetadataFeature {
     fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
-        write!(f, "MetadataFeature({self})")
+        let dbg_map: BTreeMap<&str, String> = self
```

Comment:
bit inconsistent ... here you create a binding while for `Display` you don't.

Reply:
Done